kernel/userspace.c

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+#include <kernel.h>
+#include <string.h>
+#include <misc/printk.h>
+#include <kernel_structs.h>
+#include <sys_io.h>
+
+/**
+ * Kernel object validation function
+ *
+ * Retrieve metadata for a kernel object. This function is implemented in
+ * the gperf script footer, see gen_kobject_list.py
+ *
+ * @param obj Address of kernel object to get metadata
+ * @return Kernel object's metadata, or NULL if the parameter wasn't the
+ * memory address of a kernel object
+ */
+__weak _GENERIC_SECTION(.kobject_data.text)
+struct _k_object *_k_object_find(void *obj)
+{
+	/* This is the __weak stub implementation, the version created by
+	 * gperf will replace it in build phase 2
+	 */
+	ARG_UNUSED(obj);
+	return NULL;
+}
+
+#ifdef CONFIG_PRINTK
+const char *otype_to_str(enum k_objects otype)
+{
+	switch (otype) {
+	case K_OBJ_ALERT:
+		return "k_alert";
+	case K_OBJ_DELAYED_WORK:
+		return "k_delayed_work";
+	case K_OBJ_MEM_SLAB:
+		return "k_mem_slab";
+	case K_OBJ_MSGQ:
+		return "k_msgq";
+	case K_OBJ_MUTEX:
+		return "k_mutex";
+	case K_OBJ_PIPE:
+		return "k_pipe";
+	case K_OBJ_SEM:
+		return "k_sem";
+	case K_OBJ_STACK:
+		return "k_stack";
+	case K_OBJ_THREAD:
+		return "k_thread";
+	case K_OBJ_TIMER:
+		return "k_timer";
+	case K_OBJ_WORK:
+		return "k_work";
+	case K_OBJ_WORK_Q:
+		return "k_work_q";
+	default:
+		return "?";
+	}
+}
+#endif
+
+/* Stub functions, to be filled in forthcoming patch sets */
+
+static void set_thread_perms(struct _k_object *ko, struct k_thread *thread)
+{
+	ARG_UNUSED(ko);
+	ARG_UNUSED(thread);
+
+	/* STUB */
+}
+
+
+static int test_thread_perms(struct _k_object *ko)
+{
+	ARG_UNUSED(ko);
+
+	/* STUB */
+
+	return 1;
+}
+
+static int _is_thread_user(void)
+{
+	/* STUB */
+
+	return 0;
+}
+
+void k_object_grant_access(void *object, struct k_thread *thread)
```

Comment:
Design note: would be really nice to have a way to do this statically, given that the objects and threads are static objects.  Would be even nicer to have it be implicit, along the lines of "a static thread has access to any static kernel object declared in the same C source file"

Reply:
i'd like to punt on this for now, but that could be a future enhancement once all the implications have been analyzed.

Reply:
https://jira.zephyrproject.org/browse/ZEP-2585 to look into this